README.md
# aws_ai

Flutter package to wrap Amazon artificial intelligence (AI) services, which provide flutter community developers with the ability to add intelligence to their applications through an API call to pre-trained services rather than developing and training their own models.
Amazon AI services are : 
* Amazon Rekognition : built on technology used by Amazon Prime Photos to analyze billions of images daily, is a service that makes it easy to add image analysis to your applications. With Rekognition, you can detect objects, scenes, and faces in images, as well as search and compare faces between images.
* Amazon Translate : a neural machine translation service that delivers fast, high-quality, and affordable language translation.  
* Amazon Polly (still not implemented): a service that turns text into lifelike speech. Polly lets you create applications that speak in over two dozen languages with a wide variety of natural sounding male and female voices to enable you to build entirely new categories of speech-enabled products. 
* Amazon Lex (still not implemented) : uses the same technology as Amazon Alexa to provide advanced deep learning functionalities of automatic speech recognition (ASR) and natural language understanding (NLU) to enable you to build applications with conversational interfaces, commonly called chatbots.


## Translate
Sample Code

```Flutter
import 'package:aws_ai/src/TranslateHandler.dart';

TranslateHandler translate = new TranslateHandler(accessKey, secretKey, region); 
String output = await translate.translate(Languages.auto, Languages.fr, "اسمي محمد");
```

Output will be a String contains JSON object with below format 
```json
{
   "SourceLanguageCode": "string",
   "TargetLanguageCode": "string",
   "TranslatedText": "string"
}
```
"TranslatedText" will be in UTF8 format.


## Rekognition
* #### Face comparison (compareFaces)
Sample Code

```Flutter
import 'package:aws_ai/src/RekognitionHandler.dart';

File sourceImagefile, targetImagefile; //load source and target images in those File objects
String accessKey, secretKey, region ; //load your aws account info in those variables

RekognitionHandler rekognition = new RekognitionHandler(accessKey, secretKey, region); 
Future<String> labelsArray = rekognition.compareFaces(sourceImagefile, targetImagefile);
```

Output will be a String contains JSON object with below format 
```json
{
 "FaceMatches": [
    {
       "Face": {
          "BoundingBox": {
             "Height": "number",
             "Left": "number",
             "Top": "number",
             "Width": "number"
          },
          "Confidence": "number",
          "Landmarks": [
             {
                "Type": "string",
                "X": "number",
                "Y": "number"
             }
          ],
          "Pose": {
             "Pitch": "number",
             "Roll": "number",
             "Yaw": "number"
          },
          "Quality": {
             "Brightness": "number",
             "Sharpness": "number"
          }
       },
       "Similarity": "number"
    }
 ],
 "SourceImageFace": {
    "BoundingBox": {
       "Height": "number",
       "Left": "number",
       "Top": "number",
       "Width": "number"
    },
    "Confidence": "number"
 },
 "SourceImageOrientationCorrection": "string",
 "TargetImageOrientationCorrection": "string",
 "UnmatchedFaces": [
    {
       "BoundingBox": {
          "Height": "number",
          "Left": "number",
          "Top": "number",
          "Width": "number"
       },
       "Confidence": "number",
       "Landmarks": [
          {
             "Type": "string",
             "X": "number",
             "Y": "number"
          }
       ],
       "Pose": {
          "Pitch": "number",
          "Roll": "number",
          "Yaw": "number"
       },
       "Quality": {
          "Brightness": "number",
          "Sharpness": "number"
       }
    }
 ]
}
```

* #### Facial analysis (detectFaces)
Sample Code

```Flutter
import 'package:aws_ai/src/RekognitionHandler.dart';

File sourceImagefile; //load source image in this File object
String accessKey, secretKey, region ; //load your aws account info in those variables

RekognitionHandler rekognition = new RekognitionHandler(accessKey, secretKey, region);
Future<String> labelsArray = rekognition.detectFaces(sourceImagefile);
```

Output will be a String contains JSON object with below format 
```json
{
 "FaceDetails": [
    {
       "AgeRange": {
          "High": "number",
          "Low": "number"
       },
       "Beard": {
          "Confidence": "number",
          "Value": "boolean"
       },
       "BoundingBox": {
          "Height": "number",
          "Left": "number",
          "Top": "number",
          "Width": "number"
       },
       "Confidence": "number",
       "Emotions": [
          {
             "Confidence": "number",
             "Type": "string"
          }
       ],
       "Eyeglasses": {
          "Confidence": "number",
          "Value": "boolean"
       },
       "EyesOpen": {
          "Confidence": "number",
          "Value": "boolean"
       },
       "Gender": {
          "Confidence": "number",
          "Value": "string"
       },
       "Landmarks": [
          {
             "Type": "string",
             "X": "number",
             "Y": "number"
          }
       ],
       "MouthOpen": {
          "Confidence": "number",
          "Value": "boolean"
       },
       "Mustache": {
          "Confidence": "number",
          "Value": "boolean"
       },
       "Pose": {
          "Pitch": "number",
          "Roll": "number",
          "Yaw": "number"
       },
       "Quality": {
          "Brightness": "number",
          "Sharpness": "number"
       },
       "Smile": {
          "Confidence": "number",
          "Value": "boolean"
       },
       "Sunglasses": {
          "Confidence": "number",
          "Value": "boolean"
       }
    }
 ],
 "OrientationCorrection": "string"
}
```

* #### Unsafe image detection (detectModerationLabels)
Sample Code

```Flutter
import 'package:aws_ai/src/RekognitionHandler.dart';

File sourceImagefile; //load source image in this File object
String accessKey, secretKey, region ; //load your aws account info in those variables

RekognitionHandler rekognition = new RekognitionHandler(accessKey, secretKey, region);
Future<String> labelsArray = rekognition.detectModerationLabels(sourceImagefile);
```

Output will be a String contains JSON object with below format 
```json
{
 "ModerationLabels": [
    {
       "Confidence": "number",
       "Name": "string",
       "ParentName": "string"
    }
 ]
}
```

* #### Celebrity recognition (recognizeCelebrities)
Sample Code

```Flutter
import 'package:aws_ai/src/RekognitionHandler.dart';

File sourceImagefile; //load source image in this File object
String accessKey, secretKey, region ; //load your aws account info in those variables

RekognitionHandler rekognition = new RekognitionHandler(accessKey, secretKey, region);
Future<String> labelsArray = rekognition.recognizeCelebrities(sourceImagefile);
```

Output will be a String contains JSON object with below format 
```json
{
    "CelebrityFaces": [
      {
         "Face": {
            "BoundingBox": {
               "Height": "number",
               "Left": "number",
               "Top": "number",
               "Width": "number"
            },
            "Confidence": "number",
            "Landmarks": [
               {
                  "Type": "string",
                  "X": "number",
                  "Y": "number"
               }
            ],
            "Pose": {
               "Pitch": "number",
               "Roll": "number",
               "Yaw": "number"
            },
            "Quality": {
               "Brightness": "number",
               "Sharpness": "number"
            }
         },
         "Id": "string",
         "MatchConfidence": "number",
         "Name": "string",
         "Urls": [ "string" ]
      }
    ],
    "OrientationCorrection": "string",
    "UnrecognizedFaces": [
      {
         "BoundingBox": {
            "Height": "number",
            "Left": "number",
            "Top": "number",
            "Width": "number"
         },
         "Confidence": "number",
         "Landmarks": [
            {
               "Type": "string",
               "X": "number",
               "Y": "number"
            }
         ],
         "Pose": {
            "Pitch": "number",
            "Roll": "number",
            "Yaw": "number"
         },
         "Quality": {
            "Brightness": "number",
            "Sharpness": "number"
         }
      }
    ]
    }
```

* #### Text in image (detectText)
Sample Code

```Flutter
import 'package:aws_ai/src/RekognitionHandler.dart';

File sourceImagefile; //load source image in this File object
String accessKey, secretKey, region ; //load your aws account info in those variables

RekognitionHandler rekognition = new RekognitionHandler(accessKey, secretKey, region);
Future<String> labelsArray = rekognition.detectText(sourceImagefile);
```

Output will be a String contains JSON object with below format 
```json
{
 "TextDetections": [
    {
       "Confidence": "number",
       "DetectedText": "string",
       "Geometry": {
          "BoundingBox": {
             "Height": "number",
             "Left": "number",
             "Top": "number",
             "Width": "number"
          },
          "Polygon": [
             {
                "X": "number",
                "Y": "number"
             }
          ]
       },
       "Id": "number",
       "ParentId": "number",
       "Type": "string"
    }
 ]
}
```

* #### Object and scene detection (detectLabels)
Sample Code

```Flutter
import 'package:aws_ai/src/RekognitionHandler.dart';

File sourceImagefile; //load source image in this File object
String accessKey, secretKey, region ; //load your aws account info in those variables

RekognitionHandler rekognition = new RekognitionHandler(accessKey, secretKey, region);
Future<String> labelsArray = rekognition.detectLabels(sourceImagefile);
```

Output will be a String contains JSON object with below format 
```json
{
 "Labels": [
    {
       "Confidence": "number",
       "Name": "string"
    }
 ],
 "OrientationCorrection": "string"
}
```




CHANGELOG.md
## [0.0.1] - 16 OCT 2018.
* Initial Open Source release.
* Support many AWS Rekognition APIs : 
 - compareFaces
 - detectFaces
 - detectModerationLabels
 - recognizeCelebrities
 - detectText
 - detectLabels


## [0.0.3] - 18 OCT 2018.
* Support AWS Translate API

lib/src/Signature.dart
import 'dart:convert';

import 'package:crypto/crypto.dart';
import 'package:crypto/src/digest_sink.dart';
import 'package:intl/intl.dart';

class Signature {
  Signature();

  static String generateSignature(
      String endpoint,
      String service,
      String region,
      String secretKey,
      String httpMethod,
      var now,
      String queryStringParamters,
      Map<String, String> headerParamters,
      String body) {
    if (secretKey.isEmpty) {
      throw ("No keys provided");
    }

    var amzFormatter = new DateFormat("yyyyMMdd'T'HHmmss'Z'");
    String amzDate =
        amzFormatter.format(now); // format should be '20170104T233405Z"

    var dateFormatter = new DateFormat('yyyyMMdd');
    String dateStamp = dateFormatter.format(
        now); // Date w/o time, used in credential scope. format should be "20170104"

    //************* TASK 1: CREATE A CANONICAL REQUEST *************
    //http://docs.aws.amazon.com/general/latest/gr/sigv4-create-canonical-request.html

    //Step 1 is to define the verb (GET, POST, etc.)--already done.

    //Step 2: Create canonical URI--the part of the URI from domain to query
    //string (use '/' if no path)
    String canonicalUri = '/';

    //Step 3: Create the canonical query string.
    // In case of Post, request parameters are passed in the body of the request and the query string
    //is blank.
    String canonicalQuerystring = queryStringParamters;

    //Step 4: Create the canonical headers. Header names must be trimmed
    //and lowercase, and sorted in code point order from low to high.
    //Note that there is a trailing \n.
    String canonicalHeaders = _getCanonicalFromDictionary(headerParamters);

    //Step 5: Create the list of signed headers. This lists the headers
    //in the canonical_headers list, delimited with ";" and in alpha order.
    //Note: The request can include any headers; canonical_headers and
    //signed_headers include those that you want to be included in the
    //hash of the request. "Host" and "x-amz-date" are always required.
    //For DynamoDB, content-type and x-amz-target are also required.
    String signedHeaders = _getSignedParamters(headerParamters);

    //Step 6: Create payload hash. In case of POST, the payload (body of
    //the request) contains the request parameters.
    //In case of GET body will be empty
    //String payload_hash = hashlib.sha256(utf8.encode(body)).hexdigest();
    var payloadHash = _digest(body);

    //Step 7: Combine elements to create canonical request
    String canonicalRequest = httpMethod +
        '\n' +
        canonicalUri +
        '\n' +
        canonicalQuerystring +
        '\n' +
        canonicalHeaders +
        '\n' +
        signedHeaders +
        '\n' +
        payloadHash.toString();
    var digest1 = _digest(canonicalRequest);

    //************* TASK 2: CREATE THE STRING TO SIGN*************
    //Match the algorithm to the hashing algorithm you use, either SHA-1 or
    //SHA-256 (recommended)
    String algorithm = 'AWS4-HMAC-SHA256';
    String credentialScope =
        dateStamp + '/' + region + '/' + service + '/' + 'aws4_request';
    String stringToSign = algorithm +
        '\n' +
        amzDate +
        '\n' +
        credentialScope +
        '\n' +
        digest1.toString();

    var signature;
    try {
      //************* TASK 3: CALCULATE THE SIGNATURE *************
      //Create the signing key using the function defined above.
      var signingKey = _getSignatureKey(secretKey, dateStamp, region, service);

      //Sign the string_to_sign using the signing_key
      signature = _signHex(signingKey, stringToSign);
    } catch (e) {
      print(e);
    }

    return signature.toString();
  }

  static String _getCanonicalFromDictionary(Map<String, String> dictionary) {
    //OP format: 'content-type:' + content_type + '\n' + 'host:' + host + '\n' + 'x-amz-date:' + amz_date + '\n' + 'x-amz-target:' + amz_target + '\n'
    String output = "";

    void iterateMapEntry(key, value) {
      output += key + ":" + value + '\n';
    }

    dictionary.forEach(iterateMapEntry);

    return output;
  }

  static String _getSignedParamters(Map<String, String> dictionary) {
    //OP format : 'content-type;host;x-amz-date;x-amz-target'
    String output = "";

    void iterateMapEntry(key, value) {
      output += key + ";";
    }

    dictionary.forEach(iterateMapEntry);

    output = output.substring(0, output.length - 1);

    return output;
  }

  static _sign(var key, String msg) {
    var _msg = utf8.encode(msg);

    var hmacSha256 = new Hmac(sha256, key); // HMAC-SHA256
    var digest = hmacSha256.convert(_msg);

    return digest.bytes;
  }

  static _getSignatureKey(
      String key, String dateStamp, String regionName, String serviceName) {
    var _key = utf8.encode('AWS4' + key);

    var kDate = _sign(_key, dateStamp);

    var kRegion = _sign(kDate, regionName);

    var kService = _sign(kRegion, serviceName);

    var kSigning = _sign(kService, 'aws4_request');

    return kSigning;
  }

  static _digest(String message) {
    var firstChunk = utf8.encode(message);
    var ds = new DigestSink();
    var s = sha256.startChunkedConversion(ds);
    s.add(firstChunk);
    s.close();
    return ds.value;
  }

  /*static _digestForUTF8(var firstChunk)
  {
    var ds = new DigestSink();
    var s = sha256.startChunkedConversion(ds);
    s.add(firstChunk);
    s.close();
    return ds.value;
  }*/

  static _signHex(var key, String message) {
    var _msg = utf8.encode(message);

    var hmacSha256 = new Hmac(sha256, key); // HMAC-SHA256
    var digest = hmacSha256.convert(_msg);

    return digest;
  }
}


lib/src/TranslateHandler.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:intl/intl.dart';

import 'Signature.dart';
import 'package:flutter/foundation.dart';



enum Languages {
  auto, //only for source language so Amazon auto detect the language
  en,     //English
  ar,     //Arabic
  zh,     //Chinese (Simplified)
  zh_TW,  //Chinese (Traditional) (zh-TW)
  cs,     //Czech (cs)
  fr,     //French (fr)
  de,     //German (de)
  it,     //Italian (it)
  ja,     //Japanese (ja)
  pt,     //Portuguese (pt)
  ru,     //Russian (ru)
  es,     //Spanish (es)
  tr      //Turkish (tr)
}

class TranslateHandler {
  final String _accessKey, _secretKey, _region;
  TranslateHandler(this._accessKey, this._secretKey, this._region);


  Future<String> _formatHttpRequest(String service, String amzTarget, String body) async {

    String endpoint = "https://$service.$_region.amazonaws.com/";
    String host = "rekognition.$_region.amazonaws.com";
    String httpMethod = "POST";


    var now = new DateTime.now().toUtc();
    var amzFormatter = new DateFormat("yyyyMMdd'T'HHmmss'Z'");
    String amzDate =
    amzFormatter.format(now); // format should be '20170104T233405Z"

    var dateFormatter = new DateFormat('yyyyMMdd');
    String dateStamp = dateFormatter.format(
        now); // Date w/o time, used in credential scope. format should be "20170104"

    int bodyLength = body.length;

    String queryStringParamters = "";
    Map<String, String> headerParamters = {
      "content-length": bodyLength.toString(),
      "content-type": "application/x-amz-json-1.1",
      "host": host,
      "x-amz-date": amzDate,
      "x-amz-target": amzTarget
    };

    String signature = Signature.generateSignature(
        endpoint,
        service,
        _region,
        _secretKey,
        httpMethod,
        now,
        queryStringParamters,
        headerParamters,
        body);

    String authorization =
        "AWS4-HMAC-SHA256 Credential=$_accessKey/$dateStamp/$_region/$service/aws4_request, SignedHeaders=content-length;content-type;host;x-amz-date;x-amz-target, Signature=$signature";
    headerParamters.putIfAbsent('Authorization', () => authorization);

    //String labelsArray = "";
    StringBuffer builder = new StringBuffer();
    try {
      HttpClient httpClient = new HttpClient();
      HttpClientRequest request = await httpClient.postUrl(Uri.parse(endpoint));

      request.headers.set('content-length', headerParamters['content-length']);
      request.headers.set('content-type', headerParamters['content-type']);
      request.headers.set('host', headerParamters['host']);
      request.headers.set('x-amz-date', headerParamters['x-amz-date']);
      request.headers.set('x-amz-target', headerParamters['x-amz-target']);
      request.headers.set('Authorization', headerParamters['Authorization']);

      request.write(body);

      HttpClientResponse response = await request.close();

      await for (String a in response.transform(utf8.decoder)) {
        builder.write(a);
      }
    } catch (e) {
      print(e);
    }

    return Future.value(builder.toString());
  }


  Future<String> _translate(String sourceLanguage, String targetLanguage, var text) async {

    try {
    String amzTarget = "AWSShineFrontendService_20170701.TranslateText";
    String service = "translate";

    String body =
        '{"SourceLanguageCode":"$sourceLanguage", "TargetLanguageCode":"$targetLanguage", "Text":"$text"}';

    var response = await _formatHttpRequest(service, amzTarget, body);

    return response;

    } catch (e) {
      print(e);
      return "{}";
    }

  }

  Future<String> translate(
      Languages _sourceLanguage, Languages _targetLanguage, String _text) async {

    try
    {
      if( _targetLanguage == Languages.auto )
      {
        throw("Target language can't be auto");
      }


      String sourceLanguage = _sourceLanguage != Languages.zh_TW ? describeEnum(_sourceLanguage) : "zh-TW";
      String targetLanguage = _targetLanguage != Languages.zh_TW ? describeEnum(_targetLanguage) : "zh-TW";


      var text = utf8.encode(_text);

      String response ,
          autoLanguage = sourceLanguage;

      if( (_sourceLanguage != Languages.en) && (_targetLanguage != Languages.en) )
      {
        response = await _translate(sourceLanguage, Languages.en.toString(), text);
        sourceLanguage = describeEnum(Languages.en) ;
        Map<String, dynamic> responseMap = json.decode(response);
        text = responseMap["TranslatedText"];
        autoLanguage  = responseMap["SourceLanguageCode"];
      }

      response = await _translate(sourceLanguage, targetLanguage, text);
      Map<String, dynamic> responseMap = json.decode(response);
      String translatedText = responseMap["TranslatedText"].toString();
      print(translatedText);

      String output = '{"SourceLanguageCode":"$autoLanguage", "TargetLanguageCode":"$targetLanguage", "TranslatedText": "$translatedText"}';

      return output;

    } catch (e) {
      print(e);
      return "{}";
    }

    /*
    Output will be in this format
    {
       "SourceLanguageCode": "string",
       "TargetLanguageCode": "string",
       "TranslatedText": "string"
    }
    */
  }

}

lib/src/RekognitionHandler.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:intl/intl.dart';

import 'Signature.dart';

class RekognitionHandler {
  final String _accessKey, _secretKey, _region;

  RekognitionHandler(this._accessKey, this._secretKey, this._region);

  Future<String> _rekognitionHttp(String amzTarget, String body) async {
    String endpoint = "https://rekognition.$_region.amazonaws.com/";
    String host = "rekognition.$_region.amazonaws.com";
    String httpMethod = "POST";
    String service = "rekognition";

    var now = new DateTime.now().toUtc();
    var amzFormatter = new DateFormat("yyyyMMdd'T'HHmmss'Z'");
    String amzDate =
        amzFormatter.format(now); // format should be '20170104T233405Z"

    var dateFormatter = new DateFormat('yyyyMMdd');
    String dateStamp = dateFormatter.format(
        now); // Date w/o time, used in credential scope. format should be "20170104"

    int bodyLength = body.length;

    String queryStringParamters = "";
    Map<String, String> headerParamters = {
      "content-length": bodyLength.toString(),
      "content-type": "application/x-amz-json-1.1",
      "host": host,
      "x-amz-date": amzDate,
      "x-amz-target": amzTarget
    };

    String signature = Signature.generateSignature(
        endpoint,
        service,
        _region,
        _secretKey,
        httpMethod,
        now,
        queryStringParamters,
        headerParamters,
        body);

    String authorization =
        "AWS4-HMAC-SHA256 Credential=$_accessKey/$dateStamp/$_region/$service/aws4_request, SignedHeaders=content-length;content-type;host;x-amz-date;x-amz-target, Signature=$signature";
    headerParamters.putIfAbsent('Authorization', () => authorization);

    //String labelsArray = "";
    StringBuffer builder = new StringBuffer();
    try {
      HttpClient httpClient = new HttpClient();
      HttpClientRequest request = await httpClient.postUrl(Uri.parse(endpoint));

      request.headers.set('content-length', headerParamters['content-length']);
      request.headers.set('content-type', headerParamters['content-type']);
      request.headers.set('host', headerParamters['host']);
      request.headers.set('x-amz-date', headerParamters['x-amz-date']);
      request.headers.set('x-amz-target', headerParamters['x-amz-target']);
      request.headers.set('Authorization', headerParamters['Authorization']);

      request.write(body);

      HttpClientResponse response = await request.close();

      await for (String a in response.transform(utf8.decoder)) {
        builder.write(a);
      }
    } catch (e) {
      print(e);
    }

    return Future.value(builder.toString());
  }

  Future<String> compareFaces(
      File sourceImagefile, File targetImagefile) async {
    try {
      List<int> sourceImageBytes = sourceImagefile.readAsBytesSync();
      String base64SourceImage = base64Encode(sourceImageBytes);
      List<int> targetImageBytes = targetImagefile.readAsBytesSync();
      String base64TargetImage = base64Encode(targetImageBytes);
      String body =
          '{"SourceImage":{"Bytes": "$base64SourceImage"},"TargetImage":{"Bytes": "$base64TargetImage"}}';
      String amzTarget = "RekognitionService.CompareFaces";

      String response = await _rekognitionHttp(amzTarget, body);
      return response;
    } catch (e) {
      print(e);
      return "{}";
    }

    /*
    Output will be in this format
      {
         "FaceMatches": [
            {
               "Face": {
                  "BoundingBox": {
                     "Height": number,
                     "Left": number,
                     "Top": number,
                     "Width": number
                  },
                  "Confidence": number,
                  "Landmarks": [
                     {
                        "Type": "string",
                        "X": number,
                        "Y": number
                     }
                  ],
                  "Pose": {
                     "Pitch": number,
                     "Roll": number,
                     "Yaw": number
                  },
                  "Quality": {
                     "Brightness": number,
                     "Sharpness": number
                  }
               },
               "Similarity": number
            }
         ],
         "SourceImageFace": {
            "BoundingBox": {
               "Height": number,
               "Left": number,
               "Top": number,
               "Width": number
            },
            "Confidence": number
         },
         "SourceImageOrientationCorrection": "string",
         "TargetImageOrientationCorrection": "string",
         "UnmatchedFaces": [
            {
               "BoundingBox": {
                  "Height": number,
                  "Left": number,
                  "Top": number,
                  "Width": number
               },
               "Confidence": number,
               "Landmarks": [
                  {
                     "Type": "string",
                     "X": number,
                     "Y": number
                  }
               ],
               "Pose": {
                  "Pitch": number,
                  "Roll": number,
                  "Yaw": number
               },
               "Quality": {
                  "Brightness": number,
                  "Sharpness": number
               }
            }
         ]
      }
    */
  }

  Future<String> detectFaces(File imagefile) async {
    try {
      List<int> imageBytes = imagefile.readAsBytesSync();
      String base64Image = base64Encode(imageBytes);
      String body = '{"Image":{"Bytes": "$base64Image"}}';
      String amzTarget = "RekognitionService.DetectFaces";

      String response = await _rekognitionHttp(amzTarget, body);
      return response;
    } catch (e) {
      print(e);
      return "{}";
    }

    /*
    Output will be in this format
      {
         "FaceDetails": [
            {
               "AgeRange": {
                  "High": number,
                  "Low": number
               },
               "Beard": {
                  "Confidence": number,
                  "Value": boolean
               },
               "BoundingBox": {
                  "Height": number,
                  "Left": number,
                  "Top": number,
                  "Width": number
               },
               "Confidence": number,
               "Emotions": [
                  {
                     "Confidence": number,
                     "Type": "string"
                  }
               ],
               "Eyeglasses": {
                  "Confidence": number,
                  "Value": boolean
               },
               "EyesOpen": {
                  "Confidence": number,
                  "Value": boolean
               },
               "Gender": {
                  "Confidence": number,
                  "Value": "string"
               },
               "Landmarks": [
                  {
                     "Type": "string",
                     "X": number,
                     "Y": number
                  }
               ],
               "MouthOpen": {
                  "Confidence": number,
                  "Value": boolean
               },
               "Mustache": {
                  "Confidence": number,
                  "Value": boolean
               },
               "Pose": {
                  "Pitch": number,
                  "Roll": number,
                  "Yaw": number
               },
               "Quality": {
                  "Brightness": number,
                  "Sharpness": number
               },
               "Smile": {
                  "Confidence": number,
                  "Value": boolean
               },
               "Sunglasses": {
                  "Confidence": number,
                  "Value": boolean
               }
            }
         ],
         "OrientationCorrection": "string"
      }
    */
  }

  Future<String> detectModerationLabels(File imagefile) async {
    try {
      List<int> imageBytes = imagefile.readAsBytesSync();
      String base64Image = base64Encode(imageBytes);
      String body = '{"Image":{"Bytes": "$base64Image"}}';
      String amzTarget = "RekognitionService.DetectModerationLabels";

      String response = await _rekognitionHttp(amzTarget, body);
      return response;
    } catch (e) {
      print(e);
      return "{}";
    }

    /*
    Output will be in this format
      {
         "ModerationLabels": [
            {
               "Confidence": number,
               "Name": "string",
               "ParentName": "string"
            }
         ]
      }
    */
  }

  Future<String> recognizeCelebrities(File imagefile) async {
    try {
      List<int> imageBytes = imagefile.readAsBytesSync();
      String base64Image = base64Encode(imageBytes);
      String body = '{"Image":{"Bytes": "$base64Image"}}';
      String amzTarget = "RekognitionService.RecognizeCelebrities";

      String response = await _rekognitionHttp(amzTarget, body);
      return response;
    } catch (e) {
      print(e);
      return "{}";
    }

    /*
    Output will be in this format
      {
           "CelebrityFaces": [
              {
                 "Face": {
                    "BoundingBox": {
                       "Height": number,
                       "Left": number,
                       "Top": number,
                       "Width": number
                    },
                    "Confidence": number,
                    "Landmarks": [
                       {
                          "Type": "string",
                          "X": number,
                          "Y": number
                       }
                    ],
                    "Pose": {
                       "Pitch": number,
                       "Roll": number,
                       "Yaw": number
                    },
                    "Quality": {
                       "Brightness": number,
                       "Sharpness": number
                    }
                 },
                 "Id": "string",
                 "MatchConfidence": number,
                 "Name": "string",
                 "Urls": [ "string" ]
              }
           ],
           "OrientationCorrection": "string",
           "UnrecognizedFaces": [
              {
                 "BoundingBox": {
                    "Height": number,
                    "Left": number,
                    "Top": number,
                    "Width": number
                 },
                 "Confidence": number,
                 "Landmarks": [
                    {
                       "Type": "string",
                       "X": number,
                       "Y": number
                    }
                 ],
                 "Pose": {
                    "Pitch": number,
                    "Roll": number,
                    "Yaw": number
                 },
                 "Quality": {
                    "Brightness": number,
                    "Sharpness": number
                 }
              }
           ]
        }
    */
  }

  Future<String> detectText(File imagefile) async {
    try {
      List<int> imageBytes = imagefile.readAsBytesSync();
      String base64Image = base64Encode(imageBytes);
      String body = '{"Image":{"Bytes": "$base64Image"}}';
      String amzTarget = "RekognitionService.DetectText";

      String response = await _rekognitionHttp(amzTarget, body);
      return response;
    } catch (e) {
      print(e);
      return "{}";
    }

    /*
    Output will be in this format
      {
         "TextDetections": [
            {
               "Confidence": number,
               "DetectedText": "string",
               "Geometry": {
                  "BoundingBox": {
                     "Height": number,
                     "Left": number,
                     "Top": number,
                     "Width": number
                  },
                  "Polygon": [
                     {
                        "X": number,
                        "Y": number
                     }
                  ]
               },
               "Id": number,
               "ParentId": number,
               "Type": "string"
            }
         ]
      }
    */
  }

  Future<String> detectLabels(File imagefile) async {
    try {
      List<int> imageBytes = imagefile.readAsBytesSync();
      String base64Image = base64Encode(imageBytes);
      String body = '{"Image":{"Bytes": "$base64Image"}}';
      String amzTarget = "RekognitionService.DetectLabels";

      String response = await _rekognitionHttp(amzTarget, body);

      return response;
    } catch (e) {
      print(e);
      return "{}";
    }

    /*
    Output will be in this format
      {
         "Labels": [
            {
               "Confidence": number,
               "Name": "string"
            }
         ],
         "OrientationCorrection": "string"
      }
    */
  }
}
